<a href="https://colab.research.google.com/github/KodingNextFC/FC-AI-Program/blob/main/%5BTeachers_Only%5D_Let's_Talk_About_Your_Zzzs_Part_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#1.Data Understanding

##1.1 Load the Dataset

**Dataset Columns:**

Person ID: An identifier for each individual.

Gender: The gender of the person (Male/Female).

Age: The age of the person in years.

Occupation: The occupation or profession of the person.

Sleep Duration (hours): The number of hours the person sleeps per day.

Quality of Sleep (scale: 1-10): A subjective rating of the quality of sleep, ranging from 1 to 10.

Physical Activity Level (minutes/day): The number of minutes the person engages in physical activity daily.

Stress Level (scale: 1-10): A subjective rating of the stress level experienced by the person, ranging from 1 to 10.

BMI Category: The BMI category of the person (e.g., Underweight, Normal, Overweight).

Blood Pressure (systolic/diastolic): The blood pressure measurement of the person, indicated as systolic pressure over diastolic pressure.

Heart Rate (bpm): The resting heart rate of the person in beats per minute.

Daily Steps: The number of steps the person takes per day.

Sleep Disorder: The presence or absence of a sleep disorder in the person (None, Insomnia, Sleep Apnea).


**Details about Sleep Disorder Column:**

**None**: The individual does not exhibit any specific sleep disorder.

**Insomnia**: The individual experiences difficulty falling asleep or staying asleep, leading to inadequate or poor-quality sleep.

**Sleep Apnea**: The individual suffers from pauses in breathing during sleep, resulting in disrupted sleep patterns and potential health risks.

In [ ]:
import pandas as pd
url = "https://drive.google.com/uc?export=download&id=1F7V52hhO7_INXYj-DscSKVxDm2juupn8"
df = pd.read_csv(url)
df.head(5)

Person ID Gender  Age            Occupation  Sleep Duration  \
0          1   Male   27     Software Engineer             6.1   
1          2   Male   28                Doctor             6.2   
2          3   Male   28                Doctor             6.2   
3          4   Male   28  Sales Representative             5.9   
4          5   Male   28  Sales Representative             5.9   

   Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                 6                       42             6   Overweight   
1                 6                       60             8       Normal   
2                 6                       60             8       Normal   
3                 4                       30             8        Obese   
4                 4                       30             8        Obese   

  Blood Pressure  Heart Rate  Daily Steps Sleep Disorder  
0         126/83          77         4200            NaN  
1         125/80          75        10000            NaN  
2         125/80          75        10000            NaN  
3         140/90          85         3000    Sleep Apnea  
4         140/90          85         3000    Sleep Apnea

##1.2 Features (Columns)

What are the features (columns) of the dataset?

In [ ]:
df.columns

Index(['Person ID', 'Gender', 'Age', 'Occupation', 'Sleep Duration',
       'Quality of Sleep', 'Physical Activity Level', 'Stress Level',
       'BMI Category', 'Blood Pressure', 'Heart Rate', 'Daily Steps',
       'Sleep Disorder'],
      dtype='object')

##1.3 Data Info

What is the shape of the dataset, how many rows does it have, and what are the data types of the columns?

In [ ]:
df.shape

(374, 13)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 374 entries, 0 to 373
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Person ID                374 non-null    int64  
 1   Gender                   374 non-null    object 
 2   Age                      374 non-null    int64  
 3   Occupation               374 non-null    object 
 4   Sleep Duration           374 non-null    float64
 5   Quality of Sleep         374 non-null    int64  
 6   Physical Activity Level  374 non-null    int64  
 7   Stress Level             374 non-null    int64  
 8   BMI Category             374 non-null    object 
 9   Blood Pressure           374 non-null    object 
 10  Heart Rate               374 non-null    int64  
 11  Daily Steps              374 non-null    int64  
 12  Sleep Disorder           155 non-null    object 
dtypes: float64(1), int64(7), object(5)
memory usage: 38.1+ KB


#2.Handling Missing Values

Are there any missing values in the dataset?

In [ ]:
df.isnull().sum()

Person ID                    0
Gender                       0
Age                          0
Occupation                   0
Sleep Duration               0
Quality of Sleep             0
Physical Activity Level      0
Stress Level                 0
BMI Category                 0
Blood Pressure               0
Heart Rate                   0
Daily Steps                  0
Sleep Disorder             219
dtype: int64

In [ ]:
df['Sleep Disorder'].value_counts()

Sleep Disorder
Sleep Apnea    78
Insomnia       77
Name: count, dtype: int64

The 'Sleep Disorder' column has 219 missing values, which is a significant concern. With 219 out of 374 rows missing, this represents a large portion of the data. We'll assume that the missing values are due to individuals not having any sleep disorder, and therefore, they did not provide this information. To handle this, we'll use **fillna** to replace the missing values with 'None'.

In [ ]:
df['Sleep Disorder']=df['Sleep Disorder'].fillna('None')

In [ ]:
df['Sleep Disorder'].value_counts()

Sleep Disorder
None           219
Sleep Apnea     78
Insomnia        77
Name: count, dtype: int64

#3.Feature Selection

In this step, we will select the features and the label:

*   Features: The columns that will be used to predict the label.

*   Label: The target column or class that we want to predict.

Let's start with the label. Since we want to predict the sleep disorder, the label will be 'Sleep Disorder'. The remaining columns can be used as features. However, it's better to drop unimportant columns, like 'Person ID', as they don't contribute to the prediction.

**Label** : 'Sleep Disorder'

**Features** : All column except 'Sleep Disorder' and 'Person ID'

In [ ]:
#Features
x = df.drop(columns=['Sleep Disorder', 'Person ID'])

#Label
y = df[['Sleep Disorder']]

In [ ]:
x.head()

Gender  Age            Occupation  Sleep Duration  Quality of Sleep  \
0   Male   27     Software Engineer             6.1                 6   
1   Male   28                Doctor             6.2                 6   
2   Male   28                Doctor             6.2                 6   
3   Male   28  Sales Representative             5.9                 4   
4   Male   28  Sales Representative             5.9                 4   

   Physical Activity Level  Stress Level BMI Category Blood Pressure  \
0                       42             6   Overweight         126/83   
1                       60             8       Normal         125/80   
2                       60             8       Normal         125/80   
3                       30             8        Obese         140/90   
4                       30             8        Obese         140/90   

   Heart Rate  Daily Steps  
0          77         4200  
1          75        10000  
2          75        10000  
3          85         3000  
4          85         3000

In [ ]:
y.head()

Sleep Disorder
0           None
1           None
2           None
3    Sleep Apnea
4    Sleep Apnea

#4.Data Transformation

##4.1 Reduced One-Hot Encoding

Next, we will convert the categorical data into numbers using reduced one-hot encoding. This will turn each category into a separate column with 0s and 1s, and we’ll remove the first category to avoid repeating information.

In [ ]:
x=pd.get_dummies(x,drop_first=True)

In [ ]:
x.head()

Age  Sleep Duration  Quality of Sleep  Physical Activity Level  \
0   27             6.1                 6                       42   
1   28             6.2                 6                       60   
2   28             6.2                 6                       60   
3   28             5.9                 4                       30   
4   28             5.9                 4                       30   

   Stress Level  Heart Rate  Daily Steps  Gender_Male  Occupation_Doctor  \
0             6          77         4200         True              False   
1             8          75        10000         True               True   
2             8          75        10000         True               True   
3             8          85         3000         True              False   
4             8          85         3000         True              False   

   Occupation_Engineer  ...  Blood Pressure_130/85  Blood Pressure_130/86  \
0                False  ...                  False                  False   
1                False  ...                  False                  False   
2                False  ...                  False                  False   
3                False  ...                  False                  False   
4                False  ...                  False                  False   

   Blood Pressure_131/86  Blood Pressure_132/87  Blood Pressure_135/88  \
0                  False                  False                  False   
1                  False                  False                  False   
2                  False                  False                  False   
3                  False                  False                  False   
4                  False                  False                  False   

   Blood Pressure_135/90  Blood Pressure_139/91  Blood Pressure_140/90  \
0                  False                  False                  False   
1                  False                  False                  False   
2                  False                  False                  False   
3                  False                  False                   True   
4                  False                  False                   True   

   Blood Pressure_140/95  Blood Pressure_142/92  
0                  False                  False  
1                  False                  False  
2                  False                  False  
3                  False                  False  
4                  False                  False  

[5 rows x 45 columns]

##4.2 Mapping the Label

0 : None

1 : Insomnia

2: Sleep Apnea

In [ ]:
label_mapping = {'None':0, 'Insomnia': 1, 'Sleep Apnea': 2}
y = y['Sleep Disorder'].map(label_mapping)


In [ ]:
y.head()

0    0
1    0
2    0
3    2
4    2
Name: Sleep Disorder, dtype: int64

#5.Splitting Dataset

In [ ]:
from sklearn.model_selection import train_test_split

# Splitting the dataset into training (80%) and testing (20%) sets
# The train_test_split function randomly splits your dataset into training and testing sets.
# If you don’t set random_state, it will split the data differently each time you run the code.
# Setting random_state=42 ensures that the split is always the same every time you run the code.

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)